pkg/cluster/spec/tikv.go

```diff
@@ -330,3 +338,78 @@ func addr(ins Instance) string {
 	}
 	return ins.GetHost() + ":" + strconv.Itoa(ins.GetPort())
 }
+
+func genLeaderCounter(topo *Specification, tlsCfg *tls.Config) func(string) (int, error) {
+	return func(id string) (int, error) {
+		statusAddress := ""
+		foundIds := []string{}
+		for _, kv := range topo.TiKVServers {
+			kvid := fmt.Sprintf("%s:%d", kv.Host, kv.Port)
+			if id == kvid {
+				statusAddress = fmt.Sprintf("%s:%d", kv.Host, kv.StatusPort)
+				break
+			}
+			foundIds = append(foundIds, kvid)
+		}
+		if statusAddress == "" {
+			return 0, fmt.Errorf("TiKV instance with ID %s not found, found %s", id, strings.Join(foundIds, ","))
+		}
+
+		transport, err := makeTransport(tlsCfg)
+		if err != nil {
+			return 0, err
+		}
+
+		mfChan := make(chan *dto.MetricFamily, 1024)
+		go func() {
+			addr := fmt.Sprintf("http://%s/metrics", statusAddress)
+			// XXX: https://github.com/tikv/tikv/issues/5340
+			//		Some TiKV versions don't handle https correctly
+			//      So we check if it's in that case first
+			if _, err := http.Get(addr); err != nil && tlsCfg != nil {
+				addr = fmt.Sprintf("https://%s/metrics", statusAddress)
+			}
+
+			if err := prom2json.FetchMetricFamilies(addr, mfChan, transport); err != nil {
+				log.Errorf("failed counting leader on %s (status addr %s), %v", id, addr, err)
+			}
+		}()
+
+		fms := []*prom2json.Family{}
+		for mf := range mfChan {
+			fm := prom2json.NewFamily(mf)
+			fms = append(fms, fm)
+		}
+		for _, fm := range fms {
```

Comment:
Why not merge the two iterations to avoid use a slice to save the `*prom2json.Family`?

Reply:
I'm afraid that there may be goroutine leak. eg: the loop found target metric and returns but the sender has not finish all metrics

Reply:
I think we can resolve it via:
```go
defer close(mfChan)
for {
}
```